askrene fixes and API updates

rustyrussell opened the pull request:
This contains a pile of API tweaks and updates, from @Lagrang3 and myself as we actually use and test the API.  Hopefully this makes it more useful!

Lagrang3 commented on plugins/askrene/askrene.c:
```diff
@@ -806,22 +773,15 @@ static struct command_result *json_askrene_create_channel(struct command *cmd,
 		return command_param_failed();
 
```
I would also add a fool-proof check for this type of query cannot be applied to auto chans:
```suggestion
if (strstarts(layer->name, "auto."))
		return command_fail(cmd, 
		                    JSONRPC2_INVALID_PARAMS,
				    "Command does not apply to auto layers.");
```

Lagrang3 replied:
Also do the same for `json_askrene_disable_channel`, `json_askrene_disable_node`, `json_askrene_age`, `json_askrene_inform_channel`.

rustyrussell replied:
Agreed! Indeed, this gets done automaticly with the API change.

Lagrang3 commented on plugins/askrene/askrene.c:
```diff
@@ -162,7 +141,7 @@ static struct command_result *param_reserve_path(struct command *cmd,
 		return command_fail_badparam(cmd, name, buffer, tok, "should be an array");
 
-	*path = tal(cmd, struct reserve_path);
+	*path = tal_arr(cmd, struct reserve_path, tok->size);
```
No it wasn't meant to be an array. It is a structure with an array of `scidds` and `amounts`.
In fact you dereference it below using `*` instead of `[]`.

Lagrang3 commented on plugins/askrene/askrene.c:
```diff
@@ -727,25 +712,21 @@ static struct command_result *json_askrene_unreserve(struct command *cmd,
-				    fmt_amount_msat(tmpctx, path->amounts[num]),
-				    r ? r->num_htlcs : 0,
-				    r ? fmt_amount_msat(tmpctx, r->amount) : "none");
+	for (size_t i = 0; i < tal_count(path); i++) {
```
Also here `path` is a pointer to a single path. The loop should go up to the size of the path `tal_count(path->scidds)`.

rustyrussell replied:
Yes, not sure what I was thinking there!  Removed...

Lagrang3 commented on plugins/askrene/reserve.c:
```diff
@@ -118,6 +119,29 @@ void reserve_sub(const struct reserve_htable *reserved,
+					      "short_channel_id_dir",
+					      r->scidd);
+		json_add_u64(js, "age_in_seconds",
+			     timemono_between(time_mono(), r->timestamp).ts.tv_sec);
```
Small optimization: make a single call to `time_mono()` before the loop. Use that value as the present time
and use that to compute the age.

Lagrang3 commented on tests/test_askrene.py:
```diff
@@ -7,23 +7,128 @@
+                         final_cltv=0)['probability_ppm']
+
+    # Can't remove wrong amounts: that's user error
+    with pytest.raises(RpcError, match="Unknown reservation"):
```
Yes, but if the user had had the error on the second path element then the first item in the path
does get unreserved and still I get an RPC error.
Then a subsequent correct call to unreserve will fail because it will not find the first path element.

I don't want to be too pedantic. I think we can leave as it is and in the future make a reserve/unreserve RPC
that works all or nothing.

rustyrussell replied:
Yes, it's user-error.  We shouldn't try to hard to save them...

Lagrang3 commented on plugins/askrene/layer.c:
```diff
@@ -223,48 +216,41 @@ const struct local_channel *layer_find_local_channel(const struct layer *layer,
+						 const struct amount_msat *max)
 {
-	struct constraint *c = layer_find_constraint_nonconst(layer, scidd, type);
+	struct constraint *c = layer_find_constraint_nonconst(layer, scidd);
```
Why not create a new constraint every time?
I can see that this would seem unnecessary if we have to put them together anyways when we compute the routes.

But the problem is that we combine an old constraint with a new one and use the time stamp of the youngest.
If the old constraint is putting extra information on the new constraint we might be saying that whatever the age
of that old information, now it is brand new, ie. we don't give a change to old constraints to become old we just refresh their time stamp every time we add new information on the same channel.

For example, last week I learned that this channel has max=1000sat, today I learned that max=11000sat, the
new combined constrained is still the value of the old constraint and it will have the time stamp of today. But those
bits of information where learned already a week ago and we are dragging them like if we had learned about it
today. They will never expire.

Lagrang3 replied:
Also one might argue that anyways that if you use `getroutes` you will never probe a channel with an amount
greater than the max of constraints, so there is no way we might learn max=1100sat today.
But why not probe other amounts besides what `getroutes` suggests?

rustyrussell replied:
Agreed, this was premature optimization!  I'll just accumulate them for now, and we can get smarter later if they get too unwieldy.

Lagrang3 commented on common/gossmap.h:
```diff
@@ -94,22 +94,33 @@ bool gossmap_local_addchan(struct gossmap_localmods *localmods,
-			      u32 base_fee,
-			      u32 proportional_fee,
-			      u16 delay,
-			      bool enabled,
-			      int dir)
-	NO_NULL_ARGS;
+			      const struct short_channel_id_dir *scidd,
+			      const bool *enabled,
+			      const struct amount_msat *htlc_min,
+			      const struct amount_msat *htlc_max,
+			      const struct amount_msat *base_fee,
+			      const u32 *proportional_fee,
+			      const u16 *delay);
```
This looks like something I wanted to push some time ago.

rustyrussell commented on plugins/askrene/askrene.c:
```diff
@@ -697,24 +690,16 @@ static struct command_result *json_askrene_reserve(struct command *cmd,
-				    fmt_short_channel_id_dir(tmpctx, &path->scidds[num]),
-				    fmt_amount_msat(tmpctx, path->amounts[num]),
-				    r ? fmt_amount_msat(tmpctx, r->amount) : "none");
+	for (size_t i = 0; i < tal_count(path); i++) {
```
Yes, I reworked this, since obviously it wasn't clear!

Lagrang3 commented on plugins/askrene/explain_failure.c:
```diff
@@ -0,0 +1,290 @@
+			explanation = describe_disabled(tmpctx, rq, &scidd);
+		else if (amount_msat_greater(amount, cap_msat))
+			explanation = describe_capacity(tmpctx, rq, &scidd, amount);
+		else {
+			struct amount_msat min, max;
+			get_constraints(rq, c, scidd.dir, &min, &max);
+			if (amount_msat_less(max, amount)) {
+				explanation = why_max_constrained(tmpctx, rq,
+								  &scidd, amount);
+			} else
+				continue;
+		}
```
We can get constraints above and simplify this else statement
```suggestion
		else if (amount_msat_less(max, amount)) {
				explanation = why_max_constrained(tmpctx, rq,
								  &scidd, amount);
		}
```

Lagrang3 commented on plugins/askrene/explain_failure.c:
```diff
@@ -0,0 +1,290 @@
+}
+
+/* Return description of why scidd is disabled scidd */
+static const char *describe_disabled(const tal_t *ctx,
```
There is another case when a channel is disabled because it is connected to a disabled node.

Lagrang3 commented on plugins/askrene/layer.c:
```diff
@@ -319,73 +347,81 @@ void layer_add_localmods(const struct layer *layer,
-						 0,
-						 false,
-						 !dir);
+						 &scidd,
+						 &enabled,
+						 &zero, &zero,
+						 NULL, NULL, NULL);
```
Now that we have the ability to select which fields we want to change, why are we passing `zero` values to `htlc_min` and `htlc_max`?

Lagrang3 commented on plugins/askrene/layer.c:
```diff
@@ -233,43 +244,45 @@ const struct local_channel *layer_find_local_channel(const struct layer *layer,
+	for (c = constraint_hash_getfirst(layer->constraints, scidd, &cit);
+	     c;
+	     c = constraint_hash_getnext(layer->constraints, scidd, &cit)) {
+		if (amount_msat_greater(c->min, *min))
+			*min = c->min;
+		if (amount_msat_less(c->max, *max))
+			*max = c->max;
```
I think this a good place to use `amount_msat_max` and `amount_msat_min`, for a tiny optimization.

rustyrussell commented:
Added fixes for your comments, fixed FIXME in tests (to give a nicer explanation, which we now do!) and trivial rebase on master.  Will merge once CI happy!
